### PR TITLE
Fixing SSL recovery, attempt 2

### PR DESCRIPTION
**Pull request by ant0nsc**
Ensuring that the SSL online evaluator writes the optimizer state to the checkpoint.

Please follow the guidelines for PRs contained [here](docs/pull_requests.md). Checklist:

- [x] Ensure that your PR is small, and implements one change.
- [x] Add unit tests for all functions that you introduced or modified.
- [x] Run PyCharm's code cleanup tools on your Python files.
- [ ] Link the correct GitHub issue for tracking.
- [x] Update the [Changelog](CHANGELOG.md) file: Describe your change in terms of 
Added/Changed/Removed/... in the "Upcoming" section.
- [x] When merging your PR, replace the default merge message with a description of your PR,
and if needed a motivation why that change was required.

### CODE REVIEW

**Comment by ant0nsc** [comment]
Example new BYOL run: antonsc_recovery2_1637149390_c18f9e8e in RadiomicsNN
Compare to present main: master_1632772736_ee8adce0 in RadiomicsNN

File: InnerEye/ML/SSL/lightning_containers/ssl_image_classifier.py [hunk 1]
```diff
@@ -64,7 +64,4 @@ def get_data_module(self) -> InnerEyeDataModuleTypes:
         return self.data_module
 
     def get_trainer_arguments(self) -> Dict[str, Any]:
-        trained_kwargs = {}
-        if self.is_debug_model:
-            trained_kwargs.update({"limit_train_batches": 1, "limit_val_batches": 1})
-        return trained_kwargs
+        return {}
```
Comment (mebristo) [inline InnerEye/ML/SSL/lightning_containers/ssl_image_classifier.py:67]: What is the point of this function?

File: InnerEye/ML/SSL/lightning_modules/simclr_module.py [hunk 2]
```diff
@@ -57,6 +58,17 @@ def __init__(self, encoder_name: str, dataset_name: str, use_7x7_first_conv_in_r
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.encoder(x)
 
+    def training_step(self, batch: BatchType, batch_idx: int) -> torch.Tensor:
```
Comment (mebristo) [inline InnerEye/ML/SSL/lightning_modules/simclr_module.py:61]: Should the return type here be consistent with that in line 67?

File: InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py [hunk 3]
```diff
@@ -112,16 +131,20 @@ def on_validation_batch_end(self, trainer: pl.Trainer,
...
+            old_mode = self.non_linear_evaluator.training
+            self.non_linear_evaluator.eval()
             loss = self.shared_step(batch, pl_module, is_training=False)
-            pl_module.log('ssl_online_evaluator/val/loss', loss, on_step=False, on_epoch=True, sync_dist=False)
+            log_on_epoch(pl_module, 'ssl_online_evaluator/val/loss', loss)
             for metric in self.val_metrics:
-                pl_module.log(f"ssl_online_evaluator/val/{metric.name}", metric, on_epoch=True,
-                              on_step=False)  # type: ignore
+                log_on_epoch(pl_module, f"ssl_online_evaluator/val/{metric.name}", metric)
+            self.non_linear_evaluator.train(old_mode)
```
Comment (mebristo) [inline InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py:140]: Can we update the docstring to include that the train method is called here?

File: InnerEye/ML/model_training.py [hunk 4]
```diff
@@ -163,6 +164,8 @@ def create_lightning_trainer(container: LightningContainer,
                       accelerator=accelerator,
                       plugins=plugins,
                       max_epochs=container.num_epochs,
+                      limit_train_batches=container.pl_limit_train_batches or 1.0,
```
Comment (mebristo) [inline InnerEye/ML/model_training.py:167]: is it correct that this should be 1.0 (float)? In a few other places this param is an integer

File: InnerEye/ML/SSL/lightning_containers/ssl_image_classifier.py (hunk 1 above)
Comment (ant0nsc) [reply InnerEye/ML/SSL/lightning_containers/ssl_image_classifier.py:67]: Added a comment in the code. Effectively, need to undo the get_trainer_arguments from the superclass.

File: InnerEye/ML/SSL/lightning_modules/simclr_module.py (hunk 2 above)
Comment (ant0nsc) [reply InnerEye/ML/SSL/lightning_modules/simclr_module.py:61]: yes, thanks. Removed T altogether.

File: InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py (hunk 3 above)
Comment (ant0nsc) [reply InnerEye/ML/SSL/lightning_modules/ssl_online_evaluator.py:140]: The method name may be confusing, this puts the module into either train or eval mode. I've added comments to clarify what this does.

File: InnerEye/ML/model_training.py (hunk 4 above)
Comment (ant0nsc) [reply InnerEye/ML/model_training.py:167]: I've added a comment to hopefully clarify.

File: InnerEye/ML/configs/ssl/CIFAR_SSL_configs.py [hunk 5]
```diff
@@ -15,7 +15,7 @@ class CIFAR10SimCLR(SSLContainer):
     def __init__(self) -> None:
         super().__init__(ssl_training_dataset_name=SSLDatasetName.CIFAR10,
                          linear_head_dataset_name=SSLDatasetName.CIFAR10,
-                         ssl_training_batch_size=512,
+                         ssl_training_batch_size=128,
```
Comment (Shruthi42) [inline InnerEye/ML/configs/ssl/CIFAR_SSL_configs.py:18]: Can we add a comment here saying how many GPUs we use (and from this the effective batch size)?

File: InnerEye/ML/configs/ssl/CXR_SSL_configs.py [hunk 6]
```diff
@@ -29,7 +29,7 @@ def __init__(self) -> None:
                          random_seed=1,
                          recovery_checkpoint_save_interval=200,
                          num_epochs=1000,
-                         ssl_training_batch_size=1200,
+                         ssl_training_batch_size=75,
```
Comment (Shruthi42) [inline InnerEye/ML/configs/ssl/CXR_SSL_configs.py:32]: Over here as well - can we specify the number of GPUs we use for training and effective batch size?

File: InnerEye/ML/model_training.py (hunk 4 above)
Comment (Shruthi42) [reply InnerEye/ML/model_training.py:167]: Pytorch lightning seems to allow you to set this param to 0.0, in which case we'd have different behaviour from lightning.

**Review by mebristo — APPROVED** [review]

**Review by Shruthi42 — APPROVED** [review]
